pyhf/__init__.py

```diff
@@ -180,59 +153,44 @@ def __init__(self, spec, **config_kwargs):
-                                         modpars)[0]
-            factors.append(mod_factor)
-        return tensorlib.product(factors)
-
-    def _histosysdelta(self, channel, sample, pars):
-        modifiers = [m['name'] for m in sample['modifiers']
-                if m['type'] == 'histosys']
-        stack = None
-        for m in modifiers:
-            modifier, modpars = self.config.modifier(m), pars[self.config.par_slice(m)]
-            assert int(modpars.shape[0]) == 1
-
-            # print 'MODPARS', type(modpars.data)
-            mod_delta = _hfinterp_code0(modifier.at_minus_one[channel['name']][sample['name']],
-                                        modifier.at_zero[channel['name']][sample['name']],
-                                        modifier.at_plus_one[channel['name']][sample['name']],
-                                        modpars)[0]
-            stack = tensorlib.stack([mod_delta]) if stack is None else tensorlib.stack([stack,mod_delta])
-
-        return tensorlib.sum(stack, axis=0) if stack is not None else None
-
     def expected_sample(self, channel, sample, pars):
         # for each sample the expected ocunts are
         # counts = (multiplicative factors) * (normsys multiplier) * (histsys delta + nominal hist)
         #        = f1*f2*f3*f4* nomsysfactor(nom_sys_alphas) * hist(hist_addition(histosys_alphas) + nomdata)
         # nomsysfactor(nom_sys_alphas)   = 1 + sum(interp(1, anchors[i][0], anchors[i][0], val=alpha)  for i in range(nom_sys_alphas))
         # hist_addition(histosys_alphas) = sum(interp(nombin, anchors[i][0],
         # anchors[i][0], val=alpha) for i in range(histosys_alphas))
-        nom = tensorlib.astensor(sample['data'])
-        histosys_delta = self._histosysdelta(channel, sample, pars)
+        #
+        # Formula:
+        #     \nu_{cb} (\phi_p, \alpha_p, \gamma_p) = \lambda_{cs} \gamma_{cb} \phi_{cs}(\alpha) \eta_{cs}(\alpha) \sigma_{csb}(\alpha)
+        # \gamma == statsys, shapefactor
+        # \phi == normfactor, overallsys
+        # \sigma == histosysdelta + nominal
+
+        # first, collect the factors from all modifiers
+        results = {'shapesys': [],
+                   'normfactor': [],
+                   'shapefactor': [],
+                   'histosys': [],
+                   'normsys': []}
+        for m in sample['modifiers']:
+            modifier, modpars = self.config.modifier(m['name']), pars[self.config.par_slice(m['name'])]
+            results[m['type']].append(modifier.apply(channel, sample, modpars))
+
+
+        # start building the entire set of factors
+        factors = []
+        factors += results['shapesys'] + results['normfactor'] + results['shapefactor']
+        # normsysfactor(nom_sys_alphas)   = 1 + sum(interp(1, anchors[i][0],
+        # anchors[i][0], val=alpha)  for i in range(nom_sys_alphas))
+        factors += [tensorlib.product(results['normsys'])]
 
```

Comment:
I think this could be 


```
factors += [results['normsys']]
```

instead of

```
factors += [tensorlib.product(results['normsys'])]
```

this would treat each normfactor separately than all at once, and we might save some computations since `tensorlib.product` is called at the end